### ISSUE TITLE
Unsupervised XENT loss

### ISSUE THREAD

**Issue by bkj** [williamleif/GraphSAGE#14]
Are you able to provide some explanation about the definition of the XENT loss?

```
    def _xent_loss(self, inputs1, inputs2, neg_samples, hard_neg_samples=None):

        aff = self.affinity(inputs1, inputs2)
        neg_aff = self.neg_cost(inputs1, neg_samples, hard_neg_samples)

        true_xent = tf.nn.sigmoid_cross_entropy_with_logits(
                labels=tf.ones_like(aff), logits=aff)

        negative_xent = tf.nn.sigmoid_cross_entropy_with_logits(
                labels=tf.zeros_like(neg_aff), logits=neg_aff)

        loss = tf.reduce_sum(true_xent) + 0.01*tf.reduce_sum(negative_xent)
        return loss
```

That doesn't (appear to) follow Equation 1 in the paper -- specifically, where's the 0.01 constant coming from?

**Comment by RexYing** [comment]
Hi,

The multiplier was added from another experiment recently, and is now set as option which defaults to 1. Sometimes it is helpful to down-weight the loss contribution for negative samples when there are many negative samples (e.g. 1000) used for each batch of training examples, in large scale applications.

Thanks for the catch!
Rex

**Comment by bkj** [comment]
Awesome -- thanks for the explanation

**Comment by light8lee** [comment]
Hello, I'm also confused about this loss function, it seems that the formula is 

![](https://latex.codecogs.com/gif.latex?z_u^Tz_v+log(\sigmoid(z_u^Tz_v))+z_u^Tz_{vn}+log(\sigmoid(z_u^Tz_{vn})))

, which is far away from the Eq (1). Am I missing some details?
